api/middleware.go

```diff
@@ -58,6 +58,24 @@ func (s *Server) addCORSHeaders(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// muxVarsToQueryParamMiddleware moves all mux variables to query parameter
+// values, failing with an error if a name collision with user-provided query
+// params occurs
+func muxVarsToQueryParamMiddleware(next http.Handler) http.Handler {
```

Comment:
Am not a huge fan of this. I know this gets us around having to `POST` everything. However, encoding everything and anything into the URL has it's own issues. Aside from some fringe formatting stuff that might come up I've personally hit issues with URLs being too long for servers/browsers/libs etc. Usually up to 1-2k chars is fine above that it highly depends on the implementation.

On the flip side I don't really have a cleaner way to avoid having to POST every request or add weird paths on verb/path collisions.

Reply:
After more consideration I have an observation and a consideration.
Given that this only encodes `mux` vars into the query path, length shouldn't be an issue unless the path in the first place was ridiculously long. As such I'm fine with keeping it as is.

On the other hand, I'm not sure what the utility of this is, because accessing those requires access to `http.Request` anyways, so might as well parse it out with `mux.Vars(r)` as the reverse doesn't hold (ie we dont map query params into mux vars) and have to do it anyways. The only case where I see the benefit of this if we use those same params in the `lib/http.go` client and send them over BUT in that case we need to know whether to parse from one or the other or have a param encoding that has some prefix and be able to reconstruct the original path. 

So unless we can also reverse query params back to mux vars this does not absolve us from having to `POST` params for typical `GET` routes.

Reply:
I'm a fan of this approach because it takes us from 3 places to supply param fields to 2. `mux variables, r.FormValue(), and r.Body` ->  `r.FormValue(), r.Body`. I see the mux.Router as an implementation detail the `api` package should care about, but the `lib` package shouldn't know about.

mapping to query params also gets us around `POST` only when used in conjunction with `lib.UnmarshalParams`, which uses the gorilla schema decoder. eg for 
* `GET http://localhost:2503/get/b5/world_bank_population` 
* muxVarsToQueryParamMiddleware will turn this into  `GET http://localhost:2503/save/b5/world_bank_population?peername=b5&name=world_bank_population`
* `api.refStringMiddleware` will map this to `refstring`, effectively giving us ``GET http://localhost:2503/save/b5/world_bank_population?peername=b5&name=world_bank_population&refstring=b5/world_bank_population`
* the gorilla schema decoder will map this to `lib.GetParams`, which should have a `refstring` field, making the GET request work

I think the approach is right, but the execution needs work. My suggestion would be to work on combining all middleware that maps mux variables to query params into one function, but I do think the approach of keeping the whole mux variable conversation in the `api` package is the right approach